test/Npgsql.Tests/Types/JsonDynamicTests.cs

```diff
@@ -218,15 +218,16 @@ await AssertType(
     [Test]
     public async Task Poco_polymorphic_mapping()
     {
-        // We don't yet support polymorphic deserialization for jsonb as $type does not come back as the first property.
-        // This could be fixed by detecting PolymorphicOptions types, always buffering their values and modifying the text.
-        if (IsJsonb)
-            return;
-
         var dataSourceBuilder = CreateDataSourceBuilder();
         dataSourceBuilder.EnableDynamicJson(jsonClrTypes: [typeof(WeatherForecast)]);
+        dataSourceBuilder.ConfigureJsonOptions(new () { AllowOutOfOrderMetadataProperties = true });
         await using var dataSource = dataSourceBuilder.Build();
 
+        var sql =
+            IsJsonb
+                ? """{"Date": "2019-09-01T00:00:00", "$type": "extended", "Summary": "Partly cloudy", "TemperatureC": 10, "TemperatureF": 49}"""
```

Comment:
This will be a possibly flaky assertion for the read side of the test - there's no guarantee that the string representation returned by PG will have `$type` as the 2nd property. So the PG implementation here might change, breaking our test.

It would be safer to replace the AssertType below with AssertTypeWrite and a read test, which would parse the string returned from PostgreSQL and assert on the properties without asserting on their ordering. I think it's fine if the same read assertion is performed for json as well (it's not our job to check that json preserves ordering even if it does).

Reply:
@roji there are tests above that do the same thing, albeit without a roundtrip to PG. I will try out your suggestions.

Reply:
Right - there would be a break here only if the PG implementation actually changes in a way that changes the order of JSON properties in the returned string. The point is that jsonb provides no guarantee on the ordering, so this is something which could actually happen.

Reply:
BTW I switched things around in my comment above: it's the AssertWrite which reads the string representation back from PG, so that's the one that needs to parse the JSON to assert its contents without ordering.